che-incubator/chectl
# feat: Implement version rollback for server restore

### What does this PR do?
Implements ability to restore to a different version of Che than deployed operator.
It added several new falgs to `server:restore` command:
`--rollback` - finds backup made automatically and restores Che operator to the previous version and restore installation from the backup
`--backup-cr` - restores Che from given backup described by a backup object (it includes Che version as well)
`--version` - before restoring Che installation, deploys given version of Che Operator (usually used together with `--snapshot-id`)

### Screenshot/screencast of this PR


### What issues does this PR fix or reference?
https://issues.redhat.com/browse/CRW-1497

### How to test this PR?
1. Edit latest CSV for OLM stable channel: increment version everywhere, change replace version to the one being replaced. Optionally change operator image.
2. Build test catalog source with the following command:
```sh
olm/buildAndPushBundleImages.sh -p openshift -c "stable"  -i quay.io/eclipse/eclipse-che-openshift-opm-catalog:preview -f true
```
The script should push images to your images registry account. Remember catalog source image name, for example: `docker.io/user/eclipse-che-openshift-opm-catalog:preview`
3. Open Openshift console, navigate to `Administration` -> `Cluster Settings` -> `Configuration` tab. Scroll to and click on `OperatorHub`. Switch to `Sources` tab. Click `Create CatalogSource`. Fill in the fields with `test` and put your catalog source image. Click `Create` button.
4. Create subscription for the real latest version in stable channel (but there is your fake one on top, so it is possible to update to). Example subscription yaml (change `startingCSV` to real latest stable one):
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: eclipse-che
  namespace: eclipse-che
spec:
  channel: stable
  installPlanApproval: Manual
  name: eclipse-che-preview-openshift
  source: test
  sourceNamespace: openshift-marketplace
  startingCSV: eclipse-che-preview-openshift.v7.36.1
```
5. Create a checluster object (from terminal or Openshift console).
6. Wait until Che is ready
7. Approve new installation plan: `oc get ip`, select pending to your fake version install plan and edit it: `oc edit ip <install-planname>`. Set `spec.approved` to `true`.
8. Wait unitl new Operator is deployed and Che is updated. New operator should create a new backup with name like: `backup-before-update-to-7-37-0` (the version should be your fake one).
9. Run `./debug server:restore --rollback`. Note, you have to stop in `olm.ts` `constructSubscription` function and replace `sourceName ` to your catalog source name (`test`). It might be omitted if your replace exiting catalog source image instead of adding a new one.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
